streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -798,6 +806,12 @@ void runOnce() {
                         final int committed = taskManager.maybeCommitActiveTasksPerUserRequested();
 
                         if (committed > 0) {
+                            if (eosBetaEnabled) {
+                                taskManager.commitAllActive();
+                                producer.commitTransaction();
```

Comment:
I think we should consider consolidate the logic of eos / non-EOS here as we are always committing all tasks (I have another single comment about this). More specifically:

1) from task-manager to within task, when committing we only trigger the underlying `producer.sendOffsetsToTxn`; here the producer could be per-thread or per-task, which does not matter since we can call this function multiple times within a txn. * in the future we can optimize this to only one sendOffsetsToTxn for eos-beta as well to reduce the num. requests.

2) then back to task-manager, we call `producer.commitTxn`: here for eos we would call that for a list of producers, one-per-task, while for eos-enabled we call it for a single producer.

As for other calls:

3) producer.beginTxn will be wrapped inside RecordCollector with a `txnInFlight` boolean in the tech debt cleanup, i.e. whenever we are about to send / commit we will check if this flag is already true; if not call beginTxn and set the flag.

4) producer.abortTxn is similar to commit as in 1) above.

5) producer.initTxn is trickier, as we know that we should try to delay its call as much as possible to avoid txn-id timeout (i.e. with eosNonBeta we call it after the task has been restored), from this PR it is not clear to me when we called it in eosBeta --- is it actually called any where? Since it is per-thread we'd have to call it after the initialization of the task-manager probably, or after the first rebalance completes and the restoration finishes, in either way it would be inconsistent compared with the per-task model. My hope is that with restoration optimization we would continue to invest in the txnId expiration would not be a problem, and then we can always just initTxns at the construction of task-manager for eosBeta.

With all above, we would not need to pollute the EOS handling logic inside the StreamThread class except the configuration.

Reply:
So if my understanding is correct, this refactoring would not make any sense if it's not based upon tech debt cleanup? I understand that the producer txn APIs will be hided behind record collector after cleanup, but I don't think Matthias has reviewed that PR yet, and based on current trunk putting everything into TaskManager is not obvious either IMHO. Maybe you two @guozhangwang @mjsax could have a sync on what our principle is for concurrent progress between 447 and tech debt cleanup, should it be a first-come, first-merge base, or we should base on 447 development again tech debt instead of trunk? I'm kind of concerned the current work becomes invalidated if we don't have an agreed plan here.

As for 5, the timing to call initTxn is in StreamTask creation. With 447, we could do the initTxn before rebalance kick-off since it is no longer the reliable fencing point.

Reply:
Just to clarify I'm not suggesting we should do what I suggested in this PR but just trying to get to some consensus on what would be the end state after both of them are done --- and if we agrees on that we can reduce unnecessary divergence while working in parallel.

As for the timing to call initTxn, my point is that it seems that the logic will be quite different since for eos, we call initTxn after the task has finished restoring (not at the construction of the task, as it might cause timeout and that's why we delayed it), but for eosBeta we would have to call initTxn when say the task-manager was constructed. Wanted to make sure from @mjsax  @abbccdda is that right?

Reply:
Adding a bit more on my thoughts about initTxns call: in eos it is per-task so we can delay the initTxn call as much as possible to make sure the pid does not get expired; in eosBeta however, initTxn has to be triggered at the thread-level, and suppose after that has been triggered and then a rebalance happens which assigned more tasks to you, and restoring them takes quite long time during which we would not process any active tasks and hence not calling txn APIs causing the pid to be expired -- note that since we still call poll regularly we would not be fenced on the consumer side.

It might be a big issue if we still have large restoration time practically, and in near future what I can think of for tackling it are in two folds:

1) try optimizing our restoration time so that we are confident such thing should be very rare.
2) allow active task processing at the same time while restoring others, hence keep the txns going; note that it is not always comprehensive since if all actives are restoring then it would not help.

Reply:
Looks like txn.id timeout is 1 min as of current:
```val coordinatorEpochAndMetadata = txnManager.getTransactionState(transactionalId).right.flatMap {
        case None =>
          val producerId = producerIdManager.generateProducerId()
          val createdMetadata = new TransactionMetadata(transactionalId = transactionalId,
            producerId = producerId,
            lastProducerId = RecordBatch.NO_PRODUCER_ID,
            producerEpoch = RecordBatch.NO_PRODUCER_EPOCH,
            lastProducerEpoch = RecordBatch.NO_PRODUCER_EPOCH,
            txnTimeoutMs = transactionTimeoutMs,
            state = Empty,
            topicPartitions = collection.mutable.Set.empty[TopicPartition],
            txnLastUpdateTimestamp = time.milliseconds())
          txnManager.putTransactionStateIfNotExists(createdMetadata)```
As we are also considering setting the txn timeout to 10 seconds, this might be a real problem

Reply:
With further reading, I think the timer should still be 7 days for pid expiration.

Reply:
Just to confirm with your reading, we have two expiration mechanism for transactional messaging:

1) txn.id, which is timed on the txn coordinator, and by default it is 7 days (if there's no txns sent by this txn.id) and timer starts upon init-pid call;
2) pid, which is timed on the brokers to remove the pid cache, and timer would only start when we have send data to that broker, by default it is 30 secs?
3) txn timeout, that is after the starting of a txn, i.e. after the first addPartitionForTxn request.

For the initTxn call it only affects the txn.id (aka the first one), so you think we are still good?